### DISCUSSION TITLE
Sonoff TH elite THR316/THR320  local setpoint control

### DISCUSSION THREAD

**Discussion by msjfb** [arendst/Tasmota#17614] [Q&A] [answered]
Has anyone looked at the possibility of adding the capacity to locally adjust the setpoint through either 2 external push-buttons (would need 2 GPIOs I guess) or a rotary switch (a potentiometer on a GPIO analog input?)
I looked through the code (I am no expert) and it seems that it is currently possible to support 2 sensors (I currently have a working THR320 +DS18b20 sensor on my testbench).
I assume that it would be possible to use the second display line and show the current setpoint)
I am in fact trying to assemble a 'regular' thermostat that I can control with MQTT and that can also be adjusted locally.

**Comment by barbudor** [comment]
You could probably do that using a dummy PWM/Dimmer output with proper DimmerRange
Then there are different solutions with a rotary encoder, 2 buttons or event a single button to drive that dimmer.
It's then just a matter of using a rule such as `on dimmer1#dimmer do TempSetTarget %value% endon` to use that dummy dimmer value as a setpoint

**Comment by msjfb** [comment]
After looking closely at the board, there are no free GPIOS that I could use. The easiest hack I can think of is to remove the LED1 and use that GPIO. I would define it as ADC, and apply varing voltage (0-3.3v) to vary the setpoint. On the software side, at this point I do not know how much work would be required.
Does this make sense ?

**Comment by Jason2866** [comment]
We do not discuss hardware changes/addons to power monitoring devices.
There is AC mains level on the GPIOs. That's by design of all devices with power monitoring chips on the PCB.

EDIT: I mixed the devices, the one you mentioned have no power monitoring. Hardware changes are okay, when done correctly.

**Comment by msjfb** [comment]
Software wise, the Tasmota32 section mentions : _DAC GPIOs are supported through Berry gpio module_, and that link states that only GPIO25 and 26 are supported. Is this set in stone?

**Reply by Jason2866** [reply]
> Software wise, the Tasmota32 section mentions : _DAC GPIOs are supported through Berry gpio module_, and that link states that only GPIO25 and 26 are supported. Is this set in stone?

Yes, since only a few GPIOS offers the possibility for DAC. Some are not usable together when WiFi is used.

**Reply by sfromis** [reply]
While most ESP32 digital GPIO functions are multiplexed to be available on nearly any pin, analog features are not, by hardware design. And indeed only half the ADC (not DAC) pins are available when using Wifi, which Tasmota always assumes.

**Comment by msjfb** [comment, marked as answer]
I have decided for now to drop the hardware modification route. The only possible route I saw is to remove 2 LEDs and put a pushbutton on the GPIOs to use as signal to increase/decrease the setpoint. Obviously not supported in the current software, and this level of software modification is out of my reach. So I guess I'll mark this as answered. Thanks.
I am however going to open a new one as I have been testing my THR320 D with a DS18B20 and I have a couple of issues.

**Reply by barbudor** [reply]
Tinkering hardware on commercial products is often not easy: lack of room, bad access to unused GPIOs, power supply at the bare minimum...
You could consider another ESP/Tasmota to DIY a remote control

**Comment by msjfb** [comment]
The THR320 is so close to what I have been looking for...
My next attempt is to go with an additional piece of hardware small enough to fit in a cover over the THR320 (no hardware modification of the THR320):
 Moteino Arduino clone with software that emulates a DS18B20 (this is already running and tested on the THR320)
This second sensor is recognized by the THR320, and that will be my setpoint temperature, modified by 2 pushbuttons connected to the Moteino. I am assuming at this point that I will be able to modifiy the setpoint with a rule in the THR320 that will monitor this second <temperature>.
I expect a problem with displaying both temperatures on the THR320 screen: righ now when 2 sensors are connected numbers appear on 2 lines, but they are the same value of the Sensor number 2 only...

**Reply by sfromis** [reply]
Not sure what happens, but similar issues in https://github.com/arendst/Tasmota/discussions/17676
